### ISSUE TITLE
Ingesting Opt-Out/In Keywords Data from Twilio 

### ISSUE THREAD

**Issue by HanaShiho** [episphere/connect#1104]
Our participants can opt out of our SMS messages by replying with pre-specified keywords, such as STOP. Currently, these responses are not getting pulled into our system from Twilio. As a result, our SMDB continues to send SMS messages to opted-out participants; however, they are blocked by Twilio and thus do not get sent to the opted-out participants. Twilio labels these undelivered messages as failed attempts. 

We want to ingest the STOP responses into our system so that SMDB doesn't send SMS messages to participants who opted out by replying with the pre-specified opt-out keywords. Similarly, if participants who previously opted out opt back in by replying with an opt-in keyword, this information should be updated in our system so the SMDB resumes sending SMS to these participants. I've included the list of pre-specified opt-out and opt-in keywords we currently have set up in Twilio below. 

Opt-Out: Stop, Cancel, End, Quit, Unsubscribe, No
Opt-In: Start, Unstop, Yes

**Comment by Davinkjohnson** [comment]
@we-ai Clarification on requirements for September scope:

In September the only required scope is to ingest the Opt-out flag from Twilio. Twilio will manage the keywords and Connect just needs to know that the participant has opted out. The Opt-out status should be tracked by updating the "Can we text this number?" (646873644) concept ID.

**Comment by we-ai** [comment]
PR#643 implements the requirements, with details below:
- SMS opt-out: Stop, Cancel, End, No, Quit, Unsubscribe
- SMS opt-in: Start, Unstop, Yes
- Data backfill: Update "can we text" (CID 646873644) to "no" (CID 104430631) when error code "21610" ("Attempt to send to unsubscribed recipient") is bounced back from Twilio

**Comment by HanaShiho** [comment]
Just to confirm, would the September release include ingesting both opt-out and opt-in status from Twilio or only opt-out status?

**Comment by HanaShiho** [comment]
ready for stage

**Comment by HanaShiho** [comment]
Ready for prod

**Comment by brotzmanmj** [comment]
One comment I wanted to make on this just so we're aware going forward:
This might be an edge case that will never or rarely happen, but... if we sent some an SMS and they opt out in Twilio, we will now be updating ingesting that to change the User Profile to 'no permission to text'. If the participant goes into MyConnect and changes the User Profile back to 'yes permission to text', Twilio will still have them as opted out and they won't get text messages. In other words, once they opt out on Twilio, the only way for them to opt back in is via Twilio (not via the User Profile).

**Comment by depietrodeanna** [comment]
And for completeness, to piggy back off of Michelle's comment: likely not needed given the potential rarity of this scenario, but if there is ever a case to make a change to address this gap, we could always insert a message upon a change made to the user profile permission to text from 'no' to 'yes.' Something to the effect of, "if you opted out of SMS messages by replying to one of our texts with a stop keyword, you will need to reply "start" to opt back in and ensure you start to receive texts from us again."

**Comment by we-ai** [comment]
Thank you @depietrodeanna for your comments.
I agree that we need to display a reminder when a participant change "can we test" from "no" to "yes" in User Profile.

**Comment by brotzmanmj** [comment]
Agreed and great suggestion. I'll create a separate issue for this in GitHub for a future release.
